{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "frozen-piece",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import vcf\n",
    "import h5py\n",
    "import csv\n",
    "import os\n",
    "from random import seed, random"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mineral-monthly",
   "metadata": {},
   "source": [
    "# Generate ClinVar Data for Semantic Zoom"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "peaceful-gazette",
   "metadata": {},
   "source": [
    "We will use a clinvar *.vcf file as an input to generate two following files:\n",
    "1. HiGlass' scalable BED file: Depending on the zoom levels, this file contain selective variant information.\n",
    "2. HiGlass' multivec file: This file contains the frequency of variant information for every binned region along genomic axis.\n",
    "\n",
    "These two files can be used together in a single Gosling.js track to allow semantic zooming:\n",
    "- When zoomed out, show density information\n",
    "- When zoomed in, show detailed information\n",
    "\n",
    "\n",
    "clodius aggregate bedfile --chromsizes-filename hg38.txt --delimiter $'\\t' --importance-column 6 --max-per-tile 80 clinvar_20200824.bed \n",
    "\n",
    "python manage.py ingest_tileset --filename data/clinvar_20200824_v1.beddb --filetype beddb --datatype bedlike --uid clinvar_20200824_v1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "packed-conspiracy",
   "metadata": {},
   "source": [
    "## Create The Two Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "artistic-harris",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_file = os.path.join('data', 'clinvar.vcf')\n",
    "bed_file = os.path.join('data', 'clinvar.bed')\n",
    "mv5_file = os.path.join('data', 'density.hdf5')\n",
    "chr_file = os.path.join('data', 'hg38_full.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "touched-powder",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We are interested only in the following information\n",
    "CHR_FILTER = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y']\n",
    "\n",
    "SIGNIFICANCE_FILTER = [\n",
    "    'Benign',\n",
    "    'Benign/Likely_benign',\n",
    "    'Likely_benign',\n",
    "    'Uncertain_significance',\n",
    "    'Likely_pathogenic',\n",
    "    'Pathogenic/Likely_pathogenic',\n",
    "    'Pathogenic',\n",
    "    'risk_factor',\n",
    "    'Conflicting_interpretations_of_pathogenicity'\n",
    "]\n",
    "\n",
    "GOLD_STARS_FILTER = [\n",
    "    'criteria_provided,_multiple_submitters,_no_conflicts',\n",
    "    'criteria_provided,_single_submitter',\n",
    "    'criteria_provided,_conflicting_interpretations',\n",
    "    'no_assertion_criteria_provided', \n",
    "    'reviewed_by_expert_panel', \n",
    "    'practice_guideline', \n",
    "    'no_interpretation_for_the_single_variant', \n",
    "    'no_assertion_provided'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "accomplished-grenada",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(chr_file, '\\t', header=None, names=['chr', 'length'])\n",
    "\n",
    "df = df[df.chr.isin(['chr' + c for c in CHR_FILTER])]\n",
    "\n",
    "TOTAL_CHROM_LEN = df.length.sum()\n",
    "\n",
    "# l = df[df.chr == 'chr1'].length\n",
    "# l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "operational-bryan",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'SIGNIFICANCE_FILTER' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-71608fb5f407>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     82\u001b[0m \u001b[0moutput\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[0mtest\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 84\u001b[0;31m \u001b[0mdensity_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m \u001b[0mc\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mzeros\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchr\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'chr'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlength\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSIGNIFICANCE_FILTER\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mCHR_FILTER\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     85\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mrecord\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvcf_data\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-9-71608fb5f407>\u001b[0m in \u001b[0;36m<dictcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     82\u001b[0m \u001b[0moutput\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[0mtest\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 84\u001b[0;31m \u001b[0mdensity_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m \u001b[0mc\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mzeros\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchr\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'chr'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlength\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSIGNIFICANCE_FILTER\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mCHR_FILTER\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     85\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mrecord\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvcf_data\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'SIGNIFICANCE_FILTER' is not defined"
     ]
    }
   ],
   "source": [
    "# This code is based on the following public codes:\n",
    "# https://github.com/higlass/higlass-clinvar/blob/master/scripts/extract_clinvar_data.py\n",
    "# https://github.com/higlass/higlass-sequence/blob/master/scripts/fasta_to_hdf5.py\n",
    "\n",
    "\"\"\"\n",
    "Extracted Keys\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "'AF_ESP'\n",
    "'AF_EXAC'\n",
    "'AF_TGP'\n",
    "'ALLELEID'\n",
    "'ALT'\n",
    "'CHROM'\n",
    "'CLNDISDB'\n",
    "'CLNDISDBINCL'\n",
    "'CLNDN'\n",
    "'CLNDNINCL'\n",
    "'CLNHGVS'\n",
    "'CLNREVSTAT'\n",
    "'CLNSIG'\n",
    "'CLNSIGCONF'\n",
    "'CLNSIGINCL'\n",
    "'CLNVC'\n",
    "'CLNVCSO'\n",
    "'CLNVI'\n",
    "'DBVARID'\n",
    "'FILTER_PASS'\n",
    "'GENEINFO'\n",
    "'ID'\n",
    "'MC'\n",
    "'ORIGIN'\n",
    "'POS'\n",
    "'QUAL'\n",
    "'REF'\n",
    "'RS'\n",
    "'SSR'\n",
    "'altlen'\n",
    "'is_snp'\n",
    "'numalt'\n",
    "\"\"\"\n",
    "\n",
    "vcf_data = vcf.Reader(open(vcf_file, 'r'))\n",
    "mv5_data = h5py.File(mv5_file, \"w\")\n",
    "\n",
    "seed(1)\n",
    "\n",
    "output = []\n",
    "test = []\n",
    "density_dict = { c : np.zeros((df[df.chr == ('chr' + c)].length.sum(), len(SIGNIFICANCE_FILTER))) for c in CHR_FILTER}\n",
    "\n",
    "for record in vcf_data:\n",
    "    #print(record.INFO)    \n",
    "    if not 'CLNSIG' in record.INFO:\n",
    "        continue\n",
    "\n",
    "    if not 'CLNREVSTAT' in record.INFO:\n",
    "        continue\n",
    "\n",
    "    gold_stars_str = ','.join(record.INFO['CLNREVSTAT'])\n",
    "\n",
    "\n",
    "    if not gold_stars_str in GOLD_STARS_FILTER:\n",
    "        test.append(gold_stars_str)\n",
    "        continue\n",
    "\n",
    "    significance = record.INFO['CLNSIG'][0]\n",
    "    if not significance in SIGNIFICANCE_FILTER:\n",
    "        #test.append(significance)\n",
    "        continue\n",
    "\n",
    "    gold_stars = 0\n",
    "    if gold_stars_str == 'practice_guideline':\n",
    "        gold_stars = 4\n",
    "    elif gold_stars_str == 'reviewed_by_expert_panel':\n",
    "        gold_stars = 3\n",
    "    elif gold_stars_str == 'criteria_provided,_multiple_submitters,_no_conflicts':\n",
    "        gold_stars = 2\n",
    "    elif gold_stars_str in ['criteria_provided,_single_submitter', 'criteria_provided,_conflicting_interpretations']:\n",
    "        gold_stars = 1\n",
    "\n",
    "    disease_name = record.INFO['CLNDN'][0] if \"CLNDN\" in record.INFO else \".\"\n",
    "\n",
    "    if disease_name == \"not_provided\":\n",
    "        disease_name = \".\"\n",
    "\n",
    "    if not record.CHROM in CHR_FILTER:\n",
    "        continue\n",
    "\n",
    "    #print(record)\n",
    "    \n",
    "    \"\"\"\n",
    "    BED\n",
    "    \"\"\"\n",
    "    data_clean = {\n",
    "        'chr': 'chr' + record.CHROM, \n",
    "        'start': +record.POS, \n",
    "        'end': +record.POS+1, \n",
    "        'ref': record.REF,\n",
    "        'alt': record.ALT[0].sequence if record.ALT[0] else \".\",\n",
    "        'importance': gold_stars + random(),\n",
    "        'gold_stars': gold_stars,\n",
    "        'significance': significance,\n",
    "        'significance_conf': ','.join(record.INFO['CLNSIGCONF']) if \"CLNSIGCONF\" in record.INFO else \".\",\n",
    "        'variant_type': record.INFO['CLNVC'] if \"CLNVC\" in record.INFO else \".\",\n",
    "        'origin': record.INFO['ORIGIN'][0] if \"ORIGIN\" in record.INFO else \".\",\n",
    "        'molecular_consequence': record.INFO['MC'][0] if \"MC\" in record.INFO else \".\",\n",
    "        'disease_name': disease_name,\n",
    "        'hgvs': record.INFO['CLNHGVS'][0].replace(\"%3D\", \"=\") if \"CLNHGVS\" in record.INFO else \".\",\n",
    "        #'database': record.INFO['CLNDISDB'][0] if \"CLNDISDB\" in record.INFO else \".\",        \n",
    "    }\n",
    "    \n",
    "    output.append(data_clean)\n",
    "    \n",
    "    \"\"\"\n",
    "    MV5\n",
    "    \"\"\"    \n",
    "    # column is significance\n",
    "    column_index = SIGNIFICANCE_FILTER.index(significance)\n",
    "    \n",
    "    # Add frequency by 1\n",
    "    density_dict[record.CHROM][record.POS, column_index] = density_dict[record.CHROM][record.POS, column_index] + 1\n",
    "    \n",
    "myset = set(test)\n",
    "print(myset)\n",
    "\n",
    "print(output[0])\n",
    "print(output[33])\n",
    "\n",
    "print(\"Writing to BED file\")\n",
    "headers = output[0].keys()\n",
    "with open(bed_file, 'w') as opf:\n",
    "    myWriter = csv.DictWriter(opf, delimiter='\\t', fieldnames=headers)\n",
    "    for row in output:\n",
    "        myWriter.writerow(row)\n",
    "\n",
    "print(\"Writing to hdf5 file\")\n",
    "for c in CHR_FILTER:\n",
    "    print(\"chr\" + c)\n",
    "    mv5_data.create_dataset(('chr' + c), (df[df.chr == ('chr' + c)].length.sum(), len(SIGNIFICANCE_FILTER)), \"i\", data=density_dict[c], compression='gzip')\n",
    "\n",
    "mv5_data.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hawaiian-wrestling",
   "metadata": {},
   "source": [
    "Then, run the following commands:\n",
    "\n",
    "```sh\n",
    "clodius aggregate bedfile \\\n",
    "    --chromsizes-filename hg38_full.txt \\\n",
    "    --delimiter $'\\t' \\\n",
    "    --importance-column 6 \\\n",
    "    --max-per-tile 80 \\\n",
    "    clinvar.bed\n",
    "\n",
    "clodius aggregate multivec \\\n",
    "    --chromsizes-filename hg38_full.txt \\\n",
    "    --starting-resolution 1 \\\n",
    "    density.hdf5\n",
    "\n",
    "python manage.py ingest_tileset \\\n",
    "    --filename clinvar.bed.beddb \\\n",
    "    --filetype beddb \\\n",
    "    --datatype bedlike \\\n",
    "    --uid clinvar\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "graphic-projection",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
