{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "prostate-happiness",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py\n",
    "import csv\n",
    "import os\n",
    "from random import seed, random"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "textile-jones",
   "metadata": {},
   "source": [
    "# Generate GWAS Data by Disease for Semantic Zoom"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "seventh-syndrome",
   "metadata": {},
   "source": [
    "We will use a GWAS file (`gwas_catalog_v1.0.2-associations_e104_r2021-10-06.tsv`, https://www.ebi.ac.uk/gwas/docs/file-downloads) as an input to generate two following files:\n",
    "1. HiGlass' BEDDB file: This file selectively stores variant information for given zoom level.\n",
    "2. HiGlass' multivec file: This file stores the frequency of variant information for every binned genomic regions.\n",
    "\n",
    "These two files can be used together in a single Gosling.js track to allow semantic zooming:\n",
    "- When zoomed out, show density information\n",
    "- When zoomed in, show detailed information\n",
    "\n",
    "Find information about the file header of GWAS files at https://www.ebi.ac.uk/gwas/docs/fileheaders.\n",
    "\n",
    "clodius aggregate bedfile --chromsizes-filename hg38.txt --delimiter $'\\t' --importance-column 6 --max-per-tile 80 gwas.v1.bed \n",
    "\n",
    "python manage.py ingest_tileset --filename data/gwas.v1.beddb --filetype beddb --datatype bedlike --uid gwas-v1-beddb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "oriental-projection",
   "metadata": {},
   "source": [
    "## Create The Two Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "pacific-field",
   "metadata": {},
   "outputs": [],
   "source": [
    "original_file_name = 'gwas_catalog_v1.0.2-associations_e104_r2021-10-06.tsv'\n",
    "original_file = os.path.join('data', original_file_name)\n",
    "bed_file = os.path.join('data', 'gwas.bed')\n",
    "mv5_file = os.path.join('data', 'gwas.hdf5')\n",
    "chr_file = os.path.join('data', 'hg38_full.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "roman-centre",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We are interested only in the following information\n",
    "CHR_FILTER = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "closing-allergy",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sehilyi/opt/miniconda3/envs/gosling-data/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument 'filepath_or_buffer' will be keyword-only\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(chr_file, '\\t', header=None, names=['chr', 'length'])\n",
    "\n",
    "df = df[df.chr.isin(['chr' + c for c in CHR_FILTER])]\n",
    "\n",
    "TOTAL_CHROM_LEN = df.length.sum()\n",
    "\n",
    "# l = df[df.chr == 'chr1'].length\n",
    "# l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "proper-matthew",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sehilyi/opt/miniconda3/envs/gosling-data/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: DtypeWarning: Columns (9,12,23,29) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LINK</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>CHR_ID</th>\n",
       "      <th>CHR_POS</th>\n",
       "      <th>REPORTED GENE(S)</th>\n",
       "      <th>MAPPED_GENE</th>\n",
       "      <th>SNPS</th>\n",
       "      <th>SNP_ID_CURRENT</th>\n",
       "      <th>P-VALUE</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>CNV</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>981</th>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>18</td>\n",
       "      <td>32526319</td>\n",
       "      <td>NR</td>\n",
       "      <td>WBP11P1 - NA</td>\n",
       "      <td>rs191218575</td>\n",
       "      <td>191218575</td>\n",
       "      <td>8.000000e-06</td>\n",
       "      <td>5.096910</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>N</td>\n",
       "      <td>breast carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982</th>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>10</td>\n",
       "      <td>51918266</td>\n",
       "      <td>NR</td>\n",
       "      <td>PRKG1</td>\n",
       "      <td>rs1937680</td>\n",
       "      <td>1937680</td>\n",
       "      <td>3.000000e-07</td>\n",
       "      <td>6.522879</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>N</td>\n",
       "      <td>breast carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983</th>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>1</td>\n",
       "      <td>118111143</td>\n",
       "      <td>NR</td>\n",
       "      <td>SPAG17</td>\n",
       "      <td>rs1962373</td>\n",
       "      <td>1962373</td>\n",
       "      <td>1.000000e-06</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>N</td>\n",
       "      <td>breast carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>4</td>\n",
       "      <td>175011701</td>\n",
       "      <td>NR</td>\n",
       "      <td>ADAM29 - NA</td>\n",
       "      <td>rs1966977</td>\n",
       "      <td>1966977</td>\n",
       "      <td>6.000000e-07</td>\n",
       "      <td>6.221849</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>N</td>\n",
       "      <td>breast carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985</th>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>2</td>\n",
       "      <td>25068670</td>\n",
       "      <td>NR</td>\n",
       "      <td>EFR3B</td>\n",
       "      <td>rs1971136</td>\n",
       "      <td>1971136</td>\n",
       "      <td>5.000000e-09</td>\n",
       "      <td>8.301030</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>N</td>\n",
       "      <td>breast carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     LINK  DISEASE/TRAIT CHR_ID    CHR_POS  \\\n",
       "981  www.ncbi.nlm.nih.gov/pubmed/29059683  Breast cancer     18   32526319   \n",
       "982  www.ncbi.nlm.nih.gov/pubmed/29059683  Breast cancer     10   51918266   \n",
       "983  www.ncbi.nlm.nih.gov/pubmed/29059683  Breast cancer      1  118111143   \n",
       "984  www.ncbi.nlm.nih.gov/pubmed/29059683  Breast cancer      4  175011701   \n",
       "985  www.ncbi.nlm.nih.gov/pubmed/29059683  Breast cancer      2   25068670   \n",
       "\n",
       "    REPORTED GENE(S)   MAPPED_GENE         SNPS SNP_ID_CURRENT       P-VALUE  \\\n",
       "981               NR  WBP11P1 - NA  rs191218575      191218575  8.000000e-06   \n",
       "982               NR         PRKG1    rs1937680        1937680  3.000000e-07   \n",
       "983               NR        SPAG17    rs1962373        1962373  1.000000e-06   \n",
       "984               NR   ADAM29 - NA    rs1966977        1966977  6.000000e-07   \n",
       "985               NR         EFR3B    rs1971136        1971136  5.000000e-09   \n",
       "\n",
       "     PVALUE_MLOG P-VALUE (TEXT) CNV      MAPPED_TRAIT  \n",
       "981     5.096910           (EA)   N  breast carcinoma  \n",
       "982     6.522879           (EA)   N  breast carcinoma  \n",
       "983     6.000000           (EA)   N  breast carcinoma  \n",
       "984     6.221849           (EA)   N  breast carcinoma  \n",
       "985     8.301030           (EA)   N  breast carcinoma  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "odf = pd.read_csv(original_file, '\\t')\n",
    "\n",
    "odf = odf.drop(columns=[\n",
    "    'INITIAL SAMPLE SIZE', 'JOURNAL', 'FIRST AUTHOR', 'DATE', 'PUBMEDID', 'DATE ADDED TO CATALOG', 'REGION', 'UPSTREAM_GENE_ID', 'DOWNSTREAM_GENE_ID', 'OR or BETA', '95% CI (TEXT)', 'REPLICATION SAMPLE SIZE', 'UPSTREAM_GENE_DISTANCE', 'DOWNSTREAM_GENE_DISTANCE', 'RISK ALLELE FREQUENCY', 'INTERGENIC', 'STUDY', 'STUDY ACCESSION', 'MAPPED_TRAIT_URI', 'PLATFORM [SNPS PASSING QC]', 'MERGED', 'SNP_GENE_IDS', 'STRONGEST SNP-RISK ALLELE', 'GENOTYPING TECHNOLOGY', 'CONTEXT'\n",
    "])\n",
    "\n",
    "odf = odf[\n",
    "    (odf['DISEASE/TRAIT'].apply(lambda x: x.upper() if isinstance(x, str) else '').str.contains('CANCER')) | (odf['MAPPED_TRAIT'].apply(lambda x: x.upper() if isinstance(x, str) else '').str.contains('CANCER'))\n",
    "]\n",
    "\n",
    "odf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fd59751-123e-43d4-8f6f-cc94e6fa4a12",
   "metadata": {},
   "source": [
    "# traits = odf[~odf['MAPPED_TRAIT'].isnull()]['MAPPED_TRAIT'].unique().tolist()\n",
    "\n",
    "unique_traits = [trait for trait_str in traits for trait in trait_str.split(', ')]\n",
    "unique_traits = list(set(unique_traits))\n",
    "\n",
    "unique_traits = [trait for trait in unique_traits]\n",
    "\n",
    "unique_traits.sort()\n",
    "\n",
    "print(len(unique_traits))\n",
    "unique_traits\n",
    "traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c3b0937-86a3-42cd-a836-e81a7457c47c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the data\n",
    "odf.to_csv(os.path.join('data', f'filtered_{original_file_name}'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f58bb9a-d4bb-47d9-87e4-300bc6514ac9",
   "metadata": {},
   "source": [
    "# Below Is Deprecated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "raising-shelter",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5642"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CATEGORY_FILTER = [\n",
    "    'Anthracycline-induced cardiotoxicity in early breast cancer',\n",
    "    'Invasive breast cancer',\n",
    "    'Persistent chemotherapy-induced alopecia in breast cancer',\n",
    "    'Sensory peripheral neuropathy in microtubule targeting agent-treated breast cancer',\n",
    "    'Adverse response to chemotherapy (amenorrhea) in breast cancer',\n",
    "    'Acute post-radiotherapy pain in breast cancer',\n",
    "    'BRCA1/2-negative high-risk breast cancer',\n",
    "    'BRCA1 mutation in breast cancer',\n",
    "    'BRCA2 mutation in breast cancer',\n",
    "    'Survival in breast cancer (estrogen-receptor positive)',\n",
    "    'Plasma anastrozole concentration in anastrozole-treated estrogen receptor positive breast cancer',\n",
    "    'Breast cancer specific mortality in breast cancer',\n",
    "    'Breast cancer specific mortality in estrogen receptor negative breast cancer',\n",
    "    'Breast cancer specific mortality in estrogen receptor positive breast cancer',\n",
    "    'Response to tamoxifen in oestrogen receptor positive/HER2 negative breast cancer',\n",
    "    'Disease-free survival in breast cancer',\n",
    "    'Estrone conjugates/estrone ratio in resected early stage estrogen-receptor positive breast cancer',\n",
    "    'Estrone/androstenedione ratio in resected early stage-receptor positive breast cancer',\n",
    "    'Plasma estrone levels in resected estrogen-receptor positive breast cancer',\n",
    "    'Plasma androstenedione levels in resected early stage-receptor positive breast cancer',\n",
    "    'Plasma estrone conjugates levels in resected early stage estrogen-receptor positive breast cancer',\n",
    "    'Lobular breast cancer (menopausal hormone therapy interaction)',\n",
    "    'Estrogen receptor status in HER2 negative breast cancer',\n",
    "    'Estrogen receptor status in breast cancer',\n",
    "    'Taxane-induced peripheral neuropathy in breast cancer',\n",
    "    'HER2 status in breast cancer',\n",
    "    'Survival in endocrine treated breast cancer (estrogen-receptor positive)',\n",
    "    'Response to tamoxifen in breast cancer',\n",
    "    'Estradiol plasma levels (breast cancer)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (anti-microtubule)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (cyclophosphamide+epirubicin+/-5FU)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (cyclophosphamide+doxorubicin+/-5FU)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (paclitaxel)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (docetaxel)',\n",
    "    'Response to chemotherapy in breast cancer (hypertension) (bevacizumab)',\n",
    "    'Response to chemotherapy in breast cancer hypertensive cases (cumulative dose) (bevacizumab)',\n",
    "    'Change in LVEF in response to paclitaxel and trastuzumab in HER2+ breast cancer'\n",
    "]\n",
    "\n",
    "diseases = odf[odf['DISEASE/TRAIT'].str.contains(\"breast cancer\", na=False)]['DISEASE/TRAIT'].unique().tolist()\n",
    "\n",
    "len(diseases)\n",
    "len(odf['DISEASE/TRAIT'].unique().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "young-opera",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to BED file\n",
      "Writing to hdf5 file\n",
      "chr1\n",
      "chr2\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chrX\n",
      "chrY\n"
     ]
    }
   ],
   "source": [
    "# For the extracted keys, refer to https://www.ebi.ac.uk/gwas/docs/fileheaders\n",
    "# DATE ADDED TO CATALOG\tPUBMEDID\tFIRST AUTHOR\tDATE\tJOURNAL\tLINK\tSTUDY\tDISEASE/TRAIT\tINITIAL SAMPLE SIZE\tREPLICATION SAMPLE SIZE\tREGION\tCHR_ID\tCHR_POS\tREPORTED GENE(S)\tMAPPED_GENE\tUPSTREAM_GENE_ID\tDOWNSTREAM_GENE_ID\tSNP_GENE_IDS\tUPSTREAM_GENE_DISTANCE\tDOWNSTREAM_GENE_DISTANCE\tSTRONGEST SNP-RISK ALLELE\tSNPS\tMERGED\tSNP_ID_CURRENT\tCONTEXT\tINTERGENIC\tRISK ALLELE FREQUENCY\tP-VALUE\tPVALUE_MLOG\tP-VALUE (TEXT)\tOR or BETA\t95% CI (TEXT)\tPLATFORM [SNPS PASSING QC]\tCNV\n",
    "\n",
    "mv5_data = h5py.File(mv5_file, \"w\")\n",
    "\n",
    "seed(1)\n",
    "\n",
    "output = []\n",
    "test = []\n",
    "density_dict = { c : np.zeros((df[df.chr == ('chr' + c)].length.sum(), 1)) for c in CHR_FILTER}\n",
    "\n",
    "for index, record in odf.iterrows(): # Oh, this is so wrong..\n",
    "    # print(record)\n",
    "\n",
    "    # if not record['DISEASE/TRAIT'] in CATEGORY_FILTER:\n",
    "        # continue\n",
    "\n",
    "    if not record.CHR_ID in CHR_FILTER:\n",
    "        continue\n",
    "    \n",
    "    if not str(record.CHR_POS).isnumeric():\n",
    "        continue\n",
    "        \n",
    "    #print(record)\n",
    "    \n",
    "    \"\"\"\n",
    "    BED\n",
    "    \"\"\"\n",
    "    data_clean = {\n",
    "        'chr': 'chr' + record.CHR_ID, \n",
    "        'start': int(record.CHR_POS),\n",
    "        'end': int(record.CHR_POS) + 1,\n",
    "        'pubmedid': record.PUBMEDID,\n",
    "        'date': record.DATE,\n",
    "        'link': record.LINK,\n",
    "        'pval': record['P-VALUE'],\n",
    "        'importance': 1 / (record['P-VALUE'] + 0.0001) + random(),\n",
    "        'disease': record['DISEASE/TRAIT'],\n",
    "        'pvalmlog': record['PVALUE_MLOG'],\n",
    "        'pvaltext': record['P-VALUE (TEXT)'],\n",
    "    }\n",
    "    \n",
    "    output.append(data_clean)\n",
    "    \n",
    "    \"\"\"\n",
    "    MV5\n",
    "    \"\"\"    \n",
    "    column_index = 0\n",
    "    \n",
    "    # Add frequency by 1\n",
    "    # print(record.CHR_ID, record.CHR_POS, column_index)\n",
    "    density_dict[record.CHR_ID][int(record.CHR_POS), column_index] = density_dict[record.CHR_ID][int(record.CHR_POS), column_index] + 1\n",
    "\n",
    "print(\"Writing to BED file\")\n",
    "headers = output[0].keys()\n",
    "with open(bed_file, 'w') as opf:\n",
    "    myWriter = csv.DictWriter(opf, delimiter='\\t', fieldnames=headers)\n",
    "    for row in output:\n",
    "        myWriter.writerow(row)\n",
    "\n",
    "print(\"Writing to hdf5 file\")\n",
    "for c in CHR_FILTER:\n",
    "    print(\"chr\" + c)\n",
    "    mv5_data.create_dataset(('chr' + c), (df[df.chr == ('chr' + c)].length.sum(), 1), \"i\", data=density_dict[c], compression='gzip')\n",
    "\n",
    "mv5_data.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "atmospheric-thunder",
   "metadata": {},
   "source": [
    "Then, run the following commands:\n",
    "\n",
    "```sh\n",
    "clodius aggregate bedfile \\\n",
    "    --chromsizes-filename hg38_full.txt \\\n",
    "    --delimiter $'\\t' \\\n",
    "    --importance-column 8 \\\n",
    "    --max-per-tile 160 \\\n",
    "    gwas.bed\n",
    "\n",
    "clodius aggregate multivec \\\n",
    "    --chromsizes-filename hg38_full.txt \\\n",
    "    --starting-resolution 1 \\\n",
    "    gwas.hdf5\n",
    "\n",
    "python manage.py ingest_tileset \\\n",
    "    --filename gwas.bed.beddb \\\n",
    "    --filetype beddb \\\n",
    "    --datatype bedlike \\\n",
    "    --uid gwas-beddb\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "periodic-opposition",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
