{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "forbidden-destination",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import vcf\n",
    "import h5py\n",
    "import csv\n",
    "import os\n",
    "from random import seed, random"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sharp-trout",
   "metadata": {},
   "source": [
    "# Generate GWAS Data for Semantic Zoom"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stuffed-location",
   "metadata": {},
   "source": [
    "We will use a GWAS *.tsv file (https://www.ebi.ac.uk/gwas/docs/file-downloads) as an input to generate two following files:\n",
    "1. HiGlass' BEDDB file: This file selectively stores variant information for given zoom level.\n",
    "2. HiGlass' multivec file: This file stores the frequency of variant information for every binned genomic regions.\n",
    "\n",
    "These two files can be used together in a single Gosling.js track to allow semantic zooming:\n",
    "- When zoomed out, show density information\n",
    "- When zoomed in, show detailed information\n",
    "\n",
    "\n",
    "clodius aggregate bedfile --chromsizes-filename hg38.txt --delimiter $'\\t' --importance-column 6 --max-per-tile 80 gwas.v1.bed \n",
    "\n",
    "python manage.py ingest_tileset --filename data/gwas.v1.beddb --filetype beddb --datatype bedlike --uid gwas-v1-beddb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "challenging-screen",
   "metadata": {},
   "source": [
    "## Create The Two Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "intermediate-session",
   "metadata": {},
   "outputs": [],
   "source": [
    "original_file = os.path.join('data', 'gwas_catalog_v1.0-associations_e100_r2021-03-25.tsv')\n",
    "bed_file = os.path.join('data', 'gwas.bed')\n",
    "mv5_file = os.path.join('data', 'gwas.hdf5')\n",
    "chr_file = os.path.join('data', 'hg38_full.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "complicated-belarus",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We are interested only in the following information\n",
    "CHR_FILTER = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "literary-chuck",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(chr_file, '\\t', header=None, names=['chr', 'length'])\n",
    "\n",
    "df = df[df.chr.isin(['chr' + c for c in CHR_FILTER])]\n",
    "\n",
    "TOTAL_CHROM_LEN = df.length.sum()\n",
    "\n",
    "# l = df[df.chr == 'chr1'].length\n",
    "# l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "received-promise",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/slyi/anaconda3/envs/gos/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3155: DtypeWarning: Columns (9,11,12,23) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>CONTEXT</th>\n",
       "      <th>INTERGENIC</th>\n",
       "      <th>RISK ALLELE FREQUENCY</th>\n",
       "      <th>P-VALUE</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2020-06-04</td>\n",
       "      <td>32296059</td>\n",
       "      <td>Han Y</td>\n",
       "      <td>2020-04-15</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32296059</td>\n",
       "      <td>Genome-wide analysis highlights contribution o...</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>84,492 European ancestry cases, 2,149 African ...</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.57</td>\n",
       "      <td>2.000000e-09</td>\n",
       "      <td>8.698970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NR [&gt; 9500000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2020-02-05</td>\n",
       "      <td>31666681</td>\n",
       "      <td>Zhao B</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>Mol Psychiatry</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31666681</td>\n",
       "      <td>Large-scale GWAS reveals genetic architecture ...</td>\n",
       "      <td>White matter microstructure (axial diusivities)</td>\n",
       "      <td>17,706 European ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.000000e-09</td>\n",
       "      <td>8.522879</td>\n",
       "      <td>(Inferior fronto-occipital fasciculus)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NR [8955960] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2020-02-05</td>\n",
       "      <td>31666681</td>\n",
       "      <td>Zhao B</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>Mol Psychiatry</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31666681</td>\n",
       "      <td>Large-scale GWAS reveals genetic architecture ...</td>\n",
       "      <td>White matter microstructure (axial diusivities)</td>\n",
       "      <td>17,706 European ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-08</td>\n",
       "      <td>7.698970</td>\n",
       "      <td>(Inferior fronto-occipital fasciculus)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NR [8955960] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2020-06-04</td>\n",
       "      <td>32296059</td>\n",
       "      <td>Han Y</td>\n",
       "      <td>2020-04-15</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32296059</td>\n",
       "      <td>Genome-wide analysis highlights contribution o...</td>\n",
       "      <td>Asthma</td>\n",
       "      <td>84,492 European ancestry cases, 2,149 African ...</td>\n",
       "      <td>NR</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.12</td>\n",
       "      <td>8.000000e-09</td>\n",
       "      <td>8.096910</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NR [&gt; 9500000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2020-02-05</td>\n",
       "      <td>31666681</td>\n",
       "      <td>Zhao B</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>Mol Psychiatry</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/31666681</td>\n",
       "      <td>Large-scale GWAS reveals genetic architecture ...</td>\n",
       "      <td>White matter microstructure (axial diusivities)</td>\n",
       "      <td>17,706 European ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.000000e-09</td>\n",
       "      <td>8.522879</td>\n",
       "      <td>(Inferior fronto-occipital fasciculus)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NR [8955960] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID FIRST AUTHOR        DATE         JOURNAL  \\\n",
       "0            2020-06-04  32296059        Han Y  2020-04-15      Nat Commun   \n",
       "1            2020-02-05  31666681       Zhao B  2019-10-30  Mol Psychiatry   \n",
       "2            2020-02-05  31666681       Zhao B  2019-10-30  Mol Psychiatry   \n",
       "3            2020-06-04  32296059        Han Y  2020-04-15      Nat Commun   \n",
       "4            2020-02-05  31666681       Zhao B  2019-10-30  Mol Psychiatry   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/32296059   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/31666681   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/31666681   \n",
       "3  www.ncbi.nlm.nih.gov/pubmed/32296059   \n",
       "4  www.ncbi.nlm.nih.gov/pubmed/31666681   \n",
       "\n",
       "                                               STUDY  \\\n",
       "0  Genome-wide analysis highlights contribution o...   \n",
       "1  Large-scale GWAS reveals genetic architecture ...   \n",
       "2  Large-scale GWAS reveals genetic architecture ...   \n",
       "3  Genome-wide analysis highlights contribution o...   \n",
       "4  Large-scale GWAS reveals genetic architecture ...   \n",
       "\n",
       "                                     DISEASE/TRAIT  \\\n",
       "0                                           Asthma   \n",
       "1  White matter microstructure (axial diusivities)   \n",
       "2  White matter microstructure (axial diusivities)   \n",
       "3                                           Asthma   \n",
       "4  White matter microstructure (axial diusivities)   \n",
       "\n",
       "                                 INITIAL SAMPLE SIZE REPLICATION SAMPLE SIZE  \\\n",
       "0  84,492 European ancestry cases, 2,149 African ...                      NR   \n",
       "1               17,706 European ancestry individuals                     NaN   \n",
       "2               17,706 European ancestry individuals                     NaN   \n",
       "3  84,492 European ancestry cases, 2,149 African ...                      NR   \n",
       "4               17,706 European ancestry individuals                     NaN   \n",
       "\n",
       "   ...             CONTEXT INTERGENIC RISK ALLELE FREQUENCY       P-VALUE  \\\n",
       "0  ...  synonymous_variant        0.0                  0.57  2.000000e-09   \n",
       "1  ...      intron_variant        0.0                   NaN  3.000000e-09   \n",
       "2  ...      intron_variant        0.0                   NaN  2.000000e-08   \n",
       "3  ...      intron_variant        0.0                  0.12  8.000000e-09   \n",
       "4  ...      intron_variant        0.0                   NaN  3.000000e-09   \n",
       "\n",
       "  PVALUE_MLOG                          P-VALUE (TEXT) OR or BETA  \\\n",
       "0    8.698970                                     NaN        NaN   \n",
       "1    8.522879  (Inferior fronto-occipital fasciculus)        NaN   \n",
       "2    7.698970  (Inferior fronto-occipital fasciculus)        NaN   \n",
       "3    8.096910                                     NaN        NaN   \n",
       "4    8.522879  (Inferior fronto-occipital fasciculus)        NaN   \n",
       "\n",
       "  95% CI (TEXT)  PLATFORM [SNPS PASSING QC]  CNV  \n",
       "0           NaN    NR [> 9500000] (imputed)    N  \n",
       "1           NaN      NR [8955960] (imputed)    N  \n",
       "2           NaN      NR [8955960] (imputed)    N  \n",
       "3           NaN    NR [> 9500000] (imputed)    N  \n",
       "4           NaN      NR [8955960] (imputed)    N  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "odf = pd.read_csv(original_file, '\\t')\n",
    "odf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "rolled-tsunami",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5052"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CATEGORY_FILTER = [\n",
    "    'Anthracycline-induced cardiotoxicity in early breast cancer',\n",
    "    'Invasive breast cancer',\n",
    "    'Persistent chemotherapy-induced alopecia in breast cancer',\n",
    "    'Sensory peripheral neuropathy in microtubule targeting agent-treated breast cancer',\n",
    "    'Adverse response to chemotherapy (amenorrhea) in breast cancer',\n",
    "    'Acute post-radiotherapy pain in breast cancer',\n",
    "    'BRCA1/2-negative high-risk breast cancer',\n",
    "    'BRCA1 mutation in breast cancer',\n",
    "    'BRCA2 mutation in breast cancer',\n",
    "    'Survival in breast cancer (estrogen-receptor positive)',\n",
    "    'Plasma anastrozole concentration in anastrozole-treated estrogen receptor positive breast cancer',\n",
    "    'Breast cancer specific mortality in breast cancer',\n",
    "    'Breast cancer specific mortality in estrogen receptor negative breast cancer',\n",
    "    'Breast cancer specific mortality in estrogen receptor positive breast cancer',\n",
    "    'Response to tamoxifen in oestrogen receptor positive/HER2 negative breast cancer',\n",
    "    'Disease-free survival in breast cancer',\n",
    "    'Estrone conjugates/estrone ratio in resected early stage estrogen-receptor positive breast cancer',\n",
    "    'Estrone/androstenedione ratio in resected early stage-receptor positive breast cancer',\n",
    "    'Plasma estrone levels in resected estrogen-receptor positive breast cancer',\n",
    "    'Plasma androstenedione levels in resected early stage-receptor positive breast cancer',\n",
    "    'Plasma estrone conjugates levels in resected early stage estrogen-receptor positive breast cancer',\n",
    "    'Lobular breast cancer (menopausal hormone therapy interaction)',\n",
    "    'Estrogen receptor status in HER2 negative breast cancer',\n",
    "    'Estrogen receptor status in breast cancer',\n",
    "    'Taxane-induced peripheral neuropathy in breast cancer',\n",
    "    'HER2 status in breast cancer',\n",
    "    'Survival in endocrine treated breast cancer (estrogen-receptor positive)',\n",
    "    'Response to tamoxifen in breast cancer',\n",
    "    'Estradiol plasma levels (breast cancer)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (anti-microtubule)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (cyclophosphamide+epirubicin+/-5FU)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (cyclophosphamide+doxorubicin+/-5FU)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (paclitaxel)',\n",
    "    'Adverse response to chemotherapy in breast cancer (alopecia) (docetaxel)',\n",
    "    'Response to chemotherapy in breast cancer (hypertension) (bevacizumab)',\n",
    "    'Response to chemotherapy in breast cancer hypertensive cases (cumulative dose) (bevacizumab)',\n",
    "    'Change in LVEF in response to paclitaxel and trastuzumab in HER2+ breast cancer'\n",
    "]\n",
    "\n",
    "diseases = odf[odf['DISEASE/TRAIT'].str.contains(\"breast cancer\", na=False)]['DISEASE/TRAIT'].unique().tolist()\n",
    "\n",
    "len(diseases)\n",
    "len(odf['DISEASE/TRAIT'].unique().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "comfortable-macintosh",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing to BED file\n",
      "Writing to hdf5 file\n",
      "chr1\n",
      "chr2\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chrX\n",
      "chrY\n"
     ]
    }
   ],
   "source": [
    "# For the extracted keys, refer to https://www.ebi.ac.uk/gwas/docs/fileheaders\n",
    "# DATE ADDED TO CATALOG\tPUBMEDID\tFIRST AUTHOR\tDATE\tJOURNAL\tLINK\tSTUDY\tDISEASE/TRAIT\tINITIAL SAMPLE SIZE\tREPLICATION SAMPLE SIZE\tREGION\tCHR_ID\tCHR_POS\tREPORTED GENE(S)\tMAPPED_GENE\tUPSTREAM_GENE_ID\tDOWNSTREAM_GENE_ID\tSNP_GENE_IDS\tUPSTREAM_GENE_DISTANCE\tDOWNSTREAM_GENE_DISTANCE\tSTRONGEST SNP-RISK ALLELE\tSNPS\tMERGED\tSNP_ID_CURRENT\tCONTEXT\tINTERGENIC\tRISK ALLELE FREQUENCY\tP-VALUE\tPVALUE_MLOG\tP-VALUE (TEXT)\tOR or BETA\t95% CI (TEXT)\tPLATFORM [SNPS PASSING QC]\tCNV\n",
    "\n",
    "mv5_data = h5py.File(mv5_file, \"w\")\n",
    "\n",
    "seed(1)\n",
    "\n",
    "output = []\n",
    "test = []\n",
    "density_dict = { c : np.zeros((df[df.chr == ('chr' + c)].length.sum(), 1)) for c in CHR_FILTER}\n",
    "\n",
    "for index, record in odf.iterrows(): # Oh, this is so wrong..\n",
    "    # print(record)\n",
    "\n",
    "    # if not record['DISEASE/TRAIT'] in CATEGORY_FILTER:\n",
    "        # continue\n",
    "\n",
    "    if not record.CHR_ID in CHR_FILTER:\n",
    "        continue\n",
    "    \n",
    "    if not str(record.CHR_POS).isnumeric():\n",
    "        continue\n",
    "        \n",
    "    #print(record)\n",
    "    \n",
    "    \"\"\"\n",
    "    BED\n",
    "    \"\"\"\n",
    "    data_clean = {\n",
    "        'chr': 'chr' + record.CHR_ID, \n",
    "        'start': int(record.CHR_POS),\n",
    "        'end': int(record.CHR_POS) + 1,\n",
    "        'pubmedid': record.PUBMEDID,\n",
    "        'date': record.DATE,\n",
    "        'link': record.LINK,\n",
    "        'pval': record['P-VALUE'],\n",
    "        'importance': 1 / (record['P-VALUE'] + 0.0001) + random(),\n",
    "        'disease': record['DISEASE/TRAIT'],\n",
    "        'pvalmlog': record['PVALUE_MLOG'],\n",
    "        'pvaltext': record['P-VALUE (TEXT)'],\n",
    "    }\n",
    "    \n",
    "    output.append(data_clean)\n",
    "    \n",
    "    \"\"\"\n",
    "    MV5\n",
    "    \"\"\"    \n",
    "    column_index = 0\n",
    "    \n",
    "    # Add frequency by 1\n",
    "    # print(record.CHR_ID, record.CHR_POS, column_index)\n",
    "    density_dict[record.CHR_ID][int(record.CHR_POS), column_index] = density_dict[record.CHR_ID][int(record.CHR_POS), column_index] + 1\n",
    "\n",
    "print(\"Writing to BED file\")\n",
    "headers = output[0].keys()\n",
    "with open(bed_file, 'w') as opf:\n",
    "    myWriter = csv.DictWriter(opf, delimiter='\\t', fieldnames=headers)\n",
    "    for row in output:\n",
    "        myWriter.writerow(row)\n",
    "\n",
    "print(\"Writing to hdf5 file\")\n",
    "for c in CHR_FILTER:\n",
    "    print(\"chr\" + c)\n",
    "    mv5_data.create_dataset(('chr' + c), (df[df.chr == ('chr' + c)].length.sum(), 1), \"i\", data=density_dict[c], compression='gzip')\n",
    "\n",
    "mv5_data.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "based-massachusetts",
   "metadata": {},
   "source": [
    "Then, run the following commands:\n",
    "\n",
    "```sh\n",
    "clodius aggregate bedfile \\\n",
    "    --chromsizes-filename hg38_full.txt \\\n",
    "    --delimiter $'\\t' \\\n",
    "    --importance-column 8 \\\n",
    "    --max-per-tile 160 \\\n",
    "    gwas.bed\n",
    "\n",
    "clodius aggregate multivec \\\n",
    "    --chromsizes-filename hg38_full.txt \\\n",
    "    --starting-resolution 1 \\\n",
    "    gwas.hdf5\n",
    "\n",
    "python manage.py ingest_tileset \\\n",
    "    --filename gwas.bed.beddb \\\n",
    "    --filetype beddb \\\n",
    "    --datatype bedlike \\\n",
    "    --uid gwas-beddb\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "combined-monte",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
